mm2src/mm2_main/src/lp_swap/taker_swap.rs

```diff
@@ -1567,59 +1586,55 @@ impl TakerSwap {
-                        })
-                        .await;
+        // Use existing payment or create new one
+        let transaction = match maybe_existing_payment {
+            Some(tx) => tx,
+            None => {
+                let time_lock = match std::env::var("USE_TEST_LOCKTIME") {
+                    Ok(_) => self.r().data.started_at,
+                    Err(_) => taker_payment_lock,
+                };
 
-                    match payment {
-                        Ok(t) => t,
-                        Err(err) => {
-                            return Ok((Some(TakerSwapCommand::Finish), vec![
-                                TakerSwapEvent::TakerPaymentTransactionFailed(
-                                    ERRL!("{}", err.get_plain_text_format()).into(),
-                                ),
-                            ]));
-                        },
-                    }
-                },
-            },
-            Err(e) => {
-                return Ok((Some(TakerSwapCommand::Finish), vec![
-                    TakerSwapEvent::TakerPaymentTransactionFailed(ERRL!("{}", e).into()),
-                ]))
+                let lock_duration = self.r().data.lock_duration;
+                match self
+                    .taker_coin
+                    .send_taker_payment(SendPaymentArgs {
+                        time_lock_duration: lock_duration,
+                        time_lock,
+                        other_pubkey: &*other_taker_coin_htlc_pub,
+                        secret_hash: &secret_hash.0,
+                        amount: taker_amount_decimal,
+                        swap_contract_address: &taker_coin_swap_contract_address,
+                        swap_unique_data: &unique_data,
+                        payment_instructions: &payment_instructions,
+                        watcher_reward,
+                        wait_for_confirmation_until: taker_payment_lock,
+                    })
+                    .await
+                {
+                    Ok(t) => t,
+                    Err(err) => {
+                        return Ok((Some(TakerSwapCommand::Finish), vec![
+                            TakerSwapEvent::TakerPaymentTransactionFailed(
+                                ERRL!("{}", err.get_plain_text_format()).into(),
+                            ),
+                        ]))
+                    },
+                }
             },
         };
 
+        // Create transaction identifier and prepare `TakerPaymentSent` success event
         let tx_hash = transaction.tx_hash_as_bytes();
         let tx_hex = BytesJson::from(transaction.tx_hex());
         info!("Taker payment tx hash {:02x}", tx_hash);
         let tx_ident = TransactionIdentifier {
             tx_hex: tx_hex.clone(),
             tx_hash,
         };
-
         let mut swap_events = vec![TakerSwapEvent::TakerPaymentSent(tx_ident)];
+
+        // Process watcher logic if enabled and supported by both coins
```

Comment:
`send_taker_payment` function is quite large. I suggest moving `let watcher_reward` and `if self.ctx.use_watchers()` functionality into separate functions. This should significantly improve readability.

On the maker side, the situation is better, but the `let watcher_reward` logic could also benefit from being moved into a separate function.

Legacy code refactoring is not the goal of current PR, so l will leave this decision up to you.

Reply:
I thought about refactoring this too, but didn't want to make the PR harder to review. I can refactor it once it's approved.

Reply:
Refactor done. It's just extracting code to a function but this should be better refactored in the future.